Refactor setsquare to use MVC pattern

rolandlo opened the pull request:
~This is a continuation of #3962 and is based on #4137, #4158 and #4159 from @bhennion.~
This PR refactors the setsquare. ~Currently it contains adding a compass, but that will later be moved into another PR.~
Previously it contained a compass, but that was moved into PR #4443.
I tried to follow the MVC pattern (push variant) explained by @LittleHuba in https://github.com/xournalpp/xournalpp/pull/3962#discussion_r858021356. ~Only the last two commits are relevant.~
I will explain the most relevant commits: 

1) MVC pattern
In 6efc15adb79f885b221fa7e9b353914d7c2571c2 the MVC pattern is applied:

The class `Setsquare` is the model.
The class `SetsquareView` is the view.
The class `SetsquareController` is the controller.
The class `SetsquareInputHandler` is somewhere between a view and a controller.

- The model has a `viewPool` and a `handlerPool` for pushing updates to the `SetsquareView`(s) and `SetsquareInputHandler`(s).
It stores the data for the setsquare and the stroke (near the longest side of the setsquare or from a point close to a shorter side of the setsquare to the midpoint of the longest side) and has getters and setters.

- The view only uses the model for querying the `viewPool`. It doesn't update the model and contains only drawing methods except of the methods to update itself when pushed from the model. 

- The input handler just handles input and uses the controller to manipulate the model. It also updates itself when pushed from the model.

- The controller manipulates the model (moves, rotates and scales the setsquare, creates and updates strokes) and contains more methods, which didn't fit in one of the other classes.

For pushing updates to the view and input handler, I use the `xoj::util::Listener` class; for rendering the setsquare I use the `OverlayBase` and `ToolView` classes, all by @bhennion .

There ~are still two issues~ is still one issue with the rendering:
~1) When activating the setsquare, it is not rendered immediately. It is only rendered after it has been moved/scaled or rotated. I'm not sure why that happens. I tried to notify the setsquare after constructing instances of all classes (in `Control.cpp`), but that crashes.~

~2)~ Since the rendering uses an overlay of one of the page views and the pages are drawn one after the other, when moving the setsquare towards the next page, the fraction that lands on the next page appears to be cut, as if the setsquare was gliding under the next page. It's not like that when moving the setsquare to the page before.
![SetsquareCutFromNextPage](https://user-images.githubusercontent.com/40485433/186721832-ea18453f-3677-4f53-b8d0-3c3c7c622e29.png)

2) Derive Setsquare classes from GeometryTool classes

In order to prepare for other geometry tools (like a compass or a ruler) to be added, in 714e479e1c04e618e1aed20f74701e07b206e8a8 the Setsquare classes are derived from more generic GeometryTool classes. Those classes basically contain the functionality of the Setsquare that all geometry tools will most likely share. This makes it easier to implement new geometry tools.

- `Setsquare` is derived from `GeometryTool`
- `SetsquareView` is derived from `GeometryToolView`
- `SetsquareController` is derived from `GeometryToolController`
- `SetsquareInputHandler` is derived from `GeometryToolHandler`

This strategy is demonstrated in ~commit bb512582867027fd5bb1aba64af264946e34b80f~ PR #4443  where the Compass is added. ~Although that commit won't be part of this PR, it is included so that the strategy can be checked to work.~

Any feedback is welcome. As I had no previous exposure to the MVC pattern, I find it quite tough to find the proper way for this refactoring.

~Only the two last commits are relevant:
ff9058228d7f4e27933c0b567bcb621d9360d19f is the combination of the commits in #3962 rebased on #4159. The setsquare is functionally equivalent to the version in master. 
The real work is in adfc317dcbc3e3e8f8218b77610ba504ab7229b4~

**Edit:** Updated description, described derivation from GeometryTool classes (Oct 15)
**Edit:** Updated description after moving the compass into another PR.

bhennion commented:
Thanks for the great work!
I haven't had time to check the code in details yet, but I do have input on this point:

> 2\. Since the rendering uses an overlay of one of the page views and the pages are drawn one after the other, when moving the setsquare towards the next page, the fraction that lands on the next page appears to be cut, as if the setsquare was gliding under the next page. It's not like that when moving the setsquare to the page before.

To avoid issues like this, some overlays and tools will have to be owned, not by the PageController, but by the XournalController (currently a mix of XournalView and widget/XournalWidget). The SetSquare is one of those, as are selections for instance.
So in the end, a decent DocumentView class should have its own vector of OverlayView's.

As far as the SetSquare is concerned, I think none of its related classes should be owned by a PageSomething class, and the SetSquare classes should not refer to any specific page (there are bugs on the current master branch if you try to drag the setsquare to a different page and use it there).

rolandlo commented:
Thanks for the input.

> As far as the SetSquare is concerned, I think none of its related classes should be owned by a PageSomething class, and the SetSquare classes should not refer to any specific page (there are bugs on the current master branch if you try to drag the setsquare to a different page and use it there).

I think the setsquare must always refer to a specific page. What should happen otherwise, if you change the page layout (e.g. change from a 2-row setup to a 3-row setup), if the position of the setsquare is not relative to a specific page? Currently when you activate the setsquare, it is activated on a specific page and should only be used there. If you need it on a different page, you should deactivate the setsquare and activate it on the other page.

rolandlo commented:
I managed to fix the first rendering bug (where the setsquare was invisible until it was manipulated). So everything works again, just the setsquare is overpainted on the next page. Thinking about that issue a little more, it seems to be only a question about rendering priority, not about ownership. Maybe overlays could get a rendering priority assigned, so that overlays of lower priority would only get painted after all overlays (of all pages) of higher priority were painted.

bhennion commented:
Hey @rolandlo. Could you rebase this on latest master? I don't think this conflicts with #4158 or #4159, so it could be directly based on master. This would help the reviewing.
A couple of changes happened in #4137 before it got merged, so the rebasing on master may not be trivial.

rolandlo commented:
> Hey @rolandlo. Could you rebase this on latest master? I don't think this conflicts with #4158 or #4159, so it could be directly based on master. This would help the reviewing. A couple of changes happened in #4137 before it got merged, so the rebasing on master may not be trivial.

Sure, I will rebase on master. I have progressed more and already implemented the compass. Now I know better what the setsquare code and its integration in the code base should look like. It will take some time to split everything into sensible commits and PR's.

rolandlo commented:
> Hey @rolandlo. Could you rebase this on latest master? I don't think this conflicts with #4158 or #4159, so it could be directly based on master. This would help the reviewing. A couple of changes happened in #4137 before it got merged, so the rebasing on master may not be trivial.

The rebasing is finished. If you could take a look that would be great. I have updated the description on the top to help through the most important commits.

rolandlo commented:
Currently there are regular pointers
```
    GeometryTool* geometryTool = nullptr;
    GeometryToolController* geometryToolController = nullptr;
    GeometryToolHandler* geometryToolHandler = nullptr;
```

in `XournalView.h`. These should probably be `unique_ptr`. I tried to create `unique_ptr`s for the derived classes and `std::move` them to get `unique_ptr`s for the base classes (GeometryTool classes), but got a crash when executing. Maybe I forgot to change something elsewhere.

bhennion commented on src/core/view/SetsquareView.cpp:
```diff
@@ -2,216 +2,280 @@
-#include "undo/UndoRedoHandler.h"           // for UndoRedoHandler
-#include "util/Rectangle.h"                 // for Rectangle
-#include "view/View.h"                      // for Context
+#include "gui/XournalView.h"          // for XournalView
```
```suggestion
#include "view/Repaintable.h"
```

bhennion commented on src/core/view/SetsquareView.h:
```diff
@@ -11,197 +11,104 @@
+#include "GeometryToolView.h"
 
 class Stroke;
 class XojPageView;
```
```suggestion
```

bhennion reviewed (changes requested):
Below are a couple of details I noticed while skimming through the code. I have not look at every line of code.

I do have a more important structural remark: I feel like the very existence of gui/inputdevices/GeometryToolHandler (and derived classes) is problematic.

The content of gui/inputdevices/ should be an abstraction layer and no tool implementation should ever end up in this directory.
Of course, you created those classes because you need them: the input system is inadequate and does not handle multitouch as we'd need to avoid tool-specific code.
I'm not familiar enough with the input system to come up with a better structured solution at this point. I need to think.

(Also, we have to keep in mind that GTK4 sees a lot of changes in the input mechnisms, so porting to GTK4 will someday require extensive work on the input system.)

rolandlo commented on src/core/view/GeometryToolView.cpp:
```diff
@@ -0,0 +1,43 @@
+
+void GeometryToolView::draw(cairo_t* cr) const {
+    xoj::util::CairoSaveGuard saveGuard(cr);
+    this->drawGeometryTool(cr);
```
The mask would also need resizing when the tool is rotated, right? So it would only improve performance when the tool or the page is moved. Do you think it is worth the effort?

rolandlo replied:
Sorry, it took a while, but I have figured it out now, see commit 45113f950c50791d6e759ebb6b91b3282ee96dfc.
The mask is drawn without (translation and) rotation. So the mask doesn't need redrawing when the geometry tool is rotated (or translated). It is translated and rotated right before `mask.paintTo`. @bhennion Can you take a look at it?

Commit c408a730e323b69e7dce0dafdc7629e2b923c6ff is independent of this PR. I just noticed that `zoomChanged` is fired even when the zoom has not changed at all. That meant that the mask needed resizing/redrawing when pinching to zoom with a scale increase/decrease that was below the threshold set in the preferences.

bhennion replied:
Yes, that's what I had in mind. The code in 45113f9 looks good!

About https://github.com/xournalpp/xournalpp/commit/c408a730e323b69e7dce0dafdc7629e2b923c6ff, it may be worth splitting it to a separate PR, perhaps even destined to 1.1.3.

rolandlo commented:
> I do have a more important structural remark: I feel like the very existence of gui/inputdevices/GeometryToolHandler (and derived classes) is problematic.
> 
> The content of gui/inputdevices/ should be an abstraction layer and no tool implementation should ever end up in this directory. Of course, you created those classes because you need them: the input system is inadequate and does not handle multitouch as we'd need to avoid tool-specific code. I'm not familiar enough with the input system to come up with a better structured solution at this point. I need to think.
> 

I agree that those files look misplaced. Maybe @LittleHuba has an opinion about how to deal with it.

> (Also, we have to keep in mind that GTK4 sees a lot of changes in the input mechnisms, so porting to GTK4 will someday require extensive work on the input system.)

We should probably get acquainted with the input mechanism in GTK4 as soon as possible and try to figure out what changes will be required when porting.

LittleHuba commented:
Historically the input system was only a fix to get rid of the old c-style input handling. In general I think it is simply impossible to get rid of all tool specific code in the input handling. Thus, for me the goal would be to at least separate the unspecific handling very strictly from the tool specific handling to get something like a layered system. How that would be possible, I honestly don't know yet. 
As far as I know, GTK4 goes more in line with other frameworks where each widget does it's own parsing of input and you have a bubbling down behavior. But I never looked deeper into that.

rolandlo commented:
Rebased and fixed merge conflict.

rolandlo commented:
/azp run create-installers

rolandlo commented:
Factored out everything related to the compass into PR #4443. Also rebased on master.
This feels ready now. @bhennion Can you take another look? Only the commits from "Apply reviewer suggestions" on have changed after your first review.

bhennion commented on src/core/control/GeometryToolController.cpp:
```diff
@@ -0,0 +1,118 @@
+    view->rerenderRect(rect.x, rect.y, rect.width, rect.height);
+}
+
+void GeometryToolController::addStrokeToLayer() {
```
Is the document's mutex locked when this is called? If not, then it needs to be locked from `ensureLayerExists` to `layer->addElement`

rolandlo replied:
Fine, I will lock it.

rolandlo replied:
Unfortunately the app freezes when I lock the document there.

bhennion replied:
So it means the lock is either already locked by the caller (I could not find where), or one of the functions called tries to lock the alread locked mutex. We need to figure this out.

rolandlo replied:
You're right, the already locked mutex gets locked. I will investigate who locked it.

rolandlo replied:
The mutex when locked in the `GeometryToolController` gets locked again when calling
```
undo->addUndoAction(std::make_unique<InsertUndoAction>(page, layer, stroke.get()));
```
which calls `Control::updateWindowTitle()` via `UndoRedoHandler::fireUpdateUndoRedoButtons()` and `Control::undoRedoChanged()`.

If moving the `addUndoAction` past the mutex unlocking, there is no freeze anymore.

rolandlo replied:
Done in https://github.com/xournalpp/xournalpp/pull/4211/commits/1ff09f070b02d5f44cddd3a0bffcf62e15055a04

bhennion replied:
Wow! This is a good example of bad code structure. The window title has nothing to do with the undo-redo stack. There should absolutely not be any call to `updateWindowTitle()` in `undoRedoChanged()`...

rolandlo replied:
When a document that has no changes compared to the saved version is modified, an asterix (*) is put into the window title (before the document name). When the modification is undone, the star is removed again. That's probably the reason, why the undo-redo stack was linked to the window title update.

bhennion commented on src/core/control/GeometryToolController.h:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * @brief The stroke drawn
+     */
+    Stroke* stroke = nullptr;
```
The Stroke is owned by the controller until it's added to the layer
```suggestion
    std::unique_ptr<Stroke> stroke;
```

rolandlo replied:
That looks unproblematic. I will change that.

rolandlo replied:
Done

bhennion commented on src/core/control/SetsquareController.cpp:
```diff
@@ -0,0 +1,119 @@
+        case HYPOTENUSE:
+            return utl::Point<double>(x, -y);
+        case LEFT_LEG:
+            return utl::Point<double>((y + x) / sqrt(2.), (y - x - geometryTool->getHeight()) / sqrt(2.));
+        case RIGHT_LEG:
+            return utl::Point<double>((y - x) / sqrt(2.), (y + x - geometryTool->getHeight()) / sqrt(2.));
```
Nitpick
```suggestion
            return utl::Point<double>((y + x) / std::sqrt(2.), (y - x - geometryTool->getHeight()) / std::sqrt(2.));
        case RIGHT_LEG:
            return utl::Point<double>((y - x) / std::sqrt(2.), (y + x - geometryTool->getHeight()) / std::sqrt(2.));
```

rolandlo replied:
Ok.

rolandlo replied:
Done

bhennion commented on src/core/control/SetsquareController.cpp:
```diff
@@ -0,0 +1,119 @@
+}
+
+void SetsquareController::finalizeStroke() {
+    hypotenuseMax = NAN;
+    hypotenuseMin = NAN;
```
Is this NAN value ever compared? Used in a `min` or `max`? Is so, then using `std::numeric_limits::max/lowest` seems more appropriate.

rolandlo replied:
Yes, it is compared in `SetsquareController::updateStroke`. I can change that.

rolandlo replied:
Done

bhennion commented on src/core/control/SetsquareController.cpp:
```diff
@@ -0,0 +1,119 @@
+    addStrokeToLayer();
+}
+
+auto SetsquareController::existsStroke() -> bool { return !std::isnan(hypotenuseMax) && !std::isnan(hypotenuseMin); }
```
This is weird. Isn't `return stroke` more appropriate?

rolandlo replied:
Maybe the naming is bad. You can have the `stroke` near the hypotenuse (`existsStroke`) or as a radius from a point to the middle point of the longest side (`existsRadius`)

bhennion replied:
I see. Then yes, the names could surely be improved. Something like existsEdgeStroke and existsRadialStroke maybe?

rolandlo replied:
I like the suggestion. Renaming is done.

bhennion commented on src/core/gui/XournalView.cpp:
```diff
@@ -681,28 +685,33 @@ void XournalView::repaintSelection(bool evenWithoutSelection) {
-auto XournalView::getSetsquareView() const -> SetsquareView* {
-    g_return_val_if_fail(this->widget != nullptr, nullptr);
-    g_return_val_if_fail(GTK_IS_XOURNAL(this->widget), nullptr);
+void XournalView::makeGeometryTool(std::string tool) {
```
Wouldn't an enum type be better for the parameter here?

rolandlo replied:
Yes, I will change that.

rolandlo replied:
Done

bhennion commented on src/core/gui/XournalView.cpp:
```diff
@@ -681,28 +685,33 @@ void XournalView::repaintSelection(bool evenWithoutSelection) {
+        auto setsquare = new Setsquare();
+        view->addOverlayView(std::make_unique<xoj::view::SetsquareView>(setsquare, view, zoomControl));
+        this->geometryTool = setsquare;
+        this->geometryToolController = new SetsquareController(view, setsquare);
+        this->geometryToolHandler = new SetsquareInputHandler(this, this->geometryToolController);
```
The controller should be owned by the Control instance. Not sure about the InputHandler. Also, the name `geometryToolHandler` is confusing. The distinction between controller and handler is not clear to me...

Overall, my strategy for determining "who owns what" is: ask yourself what will happen if and when we finally split PageView and XournalView into an MVC pattern. In this instance, it's not quite clear yet but it's a good thing to keep in mind.

Also, whoever owns them, they should be handled with a smart pointer.

rolandlo replied:
`geometryToolHandler` could be renamed to `geometryToolInputHandler` if that's less confusing. The input handler is just for processing input. It asks the controller to actually control (move, scale, rotate ...) the model (geometry tool).
When migrating to Gtk4 we will just to modify the input handler, not the controller.

bhennion replied:
Yes, this would be less confusing I guess.

rolandlo replied:
Renaming is done.

rolandlo replied:
> The controller should be owned by the Control instance. Not sure about the InputHandler. Also, the name `geometryToolHandler` is confusing. The distinction between controller and handler is not clear to me...
> 
> Overall, my strategy for determining "who owns what" is: ask yourself what will happen if and when we finally split PageView and XournalView into an MVC pattern. In this instance, it's not quite clear yet but it's a good thing to keep in mind.
> 
> Also, whoever owns them, they should be handled with a smart pointer.

The `geometryToolController` is now owned by the `Control` instance, the `geometryToolInputHandler` by the `InputContext` instance. That makes the most sense to me, as they are now where they are used.

rolandlo replied:
The `geometryTool` is now owned by the `Control` instance. I see no reason to pass it forth and back.

bhennion commented on src/core/gui/XournalView.h:
```diff
@@ -183,6 +185,13 @@ class XournalView: public DocumentListener, public ZoomListener {
+    /**
+     * Geometry tool, if active
+     */
+    GeometryTool* geometryTool = nullptr;
+    GeometryToolController* geometryToolController = nullptr;
+    GeometryToolHandler* geometryToolHandler = nullptr;
```
Maybe moved to other classes (see comment above)

```suggestion
    std::unique_ptr<GeometryTool> geometryTool;
    std::unique_ptr<GeometryToolController> geometryToolController;
    std::unique_ptr<GeometryToolHandler> geometryToolHandler;
```

rolandlo replied:
I run into trouble with `unique_ptr`s (related to derived classes and casting), but I can try again.

bhennion replied:
Could you describe the problem you encountered? We might be able to help.

rolandlo replied:
Strange. This time it worked at once. Can you please check?

bhennion commented on src/core/enums/ActionGroup.enum.h:
```diff
@@ -56,7 +56,7 @@ enum ActionGroup {
     GROUP_HIGHLIGHT_POSITION,
 
-    GROUP_SETSQUARE,
+    GROUP_GEOMETRY_TOOL,
```
Question: do we want to only ever have at most one geometric tool up?

rolandlo replied:
Yes, at most one geometry tool. I see no benefit in having several geometry tools up at the same time. It's only distracting. On real paper you sometimes need two tools (ruler and set square) for parallel translation, but digitally that's the easiest thing to do (via one finger touch or arrow keys).

bhennion commented on src/core/control/Control.cpp:
```diff
@@ -1352,11 +1348,11 @@ void Control::updateDeletePageButton() {
     size_t pNr = getCurrentPageNo();
-    auto setsquareView = win->getXournal()->getSetsquareView();
-    if (setsquareView && doc->indexOf(setsquareView->getPage()) == pNr) {
-        win->getXournal()->resetSetsquareView();
+    auto geometryToolController = win->getXournal()->getGeometryToolController();
+    if (geometryToolController && doc->indexOf(geometryToolController->getPage()) == pNr) {
```
Not sure, but we may want the lock the document mutex for that

rolandlo replied:
Fine

rolandlo replied:
Done

bhennion reviewed (commented):
Some partial review. I only reviewed 11/28 files for now.

One question: in the future, do we plan to have several geometric tools displayed at the same time?

rolandlo commented:
> Some partial review. I only reviewed 11/28 files for now.
> 
> One question: in the future, do we plan to have several geometric tools displayed at the same time?

No, at least I don't. It's less distracting to switch (with one click).

rolandlo commented:
@bhennion I think I have applied all changes you requested. It cleared up quite some things. Thanks! I'll wait for the rest of your review now.

bhennion commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+    }
+    const auto page = controller->getPage();
+
+    const auto l = page->getSelectedLayer();
+    this->lines.clear();
+    for (const auto& e: l->getElements()) {
+        if (e->getType() == ELEMENT_STROKE) {
+            auto s = dynamic_cast<Stroke*>(e);
+            if (s->getPointCount() == 2) {
+                lines.push_back(s);
+            }
+        }
+    }
```
I find this weird (even if the above comment about using `sequenceStart()` is resolved).

Filtering the entire list of elements everytime the user puts a finger on the screen is bad, performance-wise. I'm not sure how to handle the issue.

[ Also, for any filtering like that, the document's mutex needs to be locked: if an element is added to the layer by another thread, all the iterators could be invalidated, leading to a segfault ]

rolandlo replied:
I added the document locking. 
Performance-wise I think there would have to be million or billions of elements on the layer to notice any difference. Under such circumstances other operations (like erasing) might become the bigger problem.
Whilst it would be possible to cache these lines and update the cache whenever needed, I think that would be more error prone and maybe not even be more performant.
Also note that users don't put a new finger on the screen every millisecond or so.

bhennion commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+    double minDist = SNAPPING_DISTANCE_TOLERANCE;
+    const auto angleSetsquare = rotation;
+    double diffAngle{NAN};
+    for (const auto& l: lines) {
+        auto first = l->getPoint(0);
+        auto second = l->getPoint(1);
+        auto dist = Snapping::distanceLine(pos, first, second);
+        auto angleLine = std::atan2(second.y - first.y, second.x - first.x);
+        auto diff = std::remainder(angleLine - angleSetsquare, M_PI_2);
```
I think to many `auto`'s make the code harder to understand.

rolandlo replied:
Ok, I replaced some `auto`s throughout this file.

bhennion commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+        this->canBlockZoom = false;
+    }
+
+    const auto center = (this->priLastPageRel + this->secLastPageRel) / 2;
+    const auto shift = this->secLastPageRel - this->priLastPageRel;
+    const auto angle = atan2(shift.y, shift.x);
+
+    const auto offset = center - lastZoomScrollCenter;
+    controller->move(offset.x, offset.y);
+    const auto angleIncrease = angle - lastAngle;
+    const auto centerRel = (this->priLastPageRel + this->secLastPageRel) / 2.0;
```
Same auto concern here

rolandlo replied:
Ok

bhennion reviewed (changes requested):
Here are some more comments.
Mostly nitpicks, except my concerns about `GeometryToolInputHandler::sequenceStart()`. See the two related comments below.

rolandlo commented on src/core/control/GeometryToolController.h:
```diff
@@ -0,0 +1,102 @@
+    void scale(double f, double cx, double cy);
+
+    /**
+     * @brief marks a point with a "x"
```
A picture says more than words:
![grafik](https://user-images.githubusercontent.com/40485433/205459403-55a76a0d-6979-4289-bc42-893f1b6790b0.png)
Here the setsquare origin is marked. It can be useful for geometric constructions to mark the rotation center, even more so for the compass in the future.

rolandlo replied:
Hopefully it's clear now.

rolandlo commented on src/core/control/GeometryToolController.h:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * @brief the page with respect to which the setsquare is initialized
+     */
+    PageRef getPage() const;
```
I tried to change that, but it resulted in a compiler warning and (when executing) in a crash.

```term
 warning: returning reference to temporary [-Wreturn-local-addr]
  119 | auto GeometryToolController::getPage() const -> const PageRef& { return view->getPage(); }
```

bhennion replied:
Ah right. PageView::getPage also returns a PageRef and not a reference to it. That's alright, just disregard my comment (or change PageView::getPage as well).

rolandlo replied:
I disregarded your comment. Didn't want to change stuff in other files, too.

rolandlo commented on src/core/control/SetsquareController.cpp:
```diff
@@ -0,0 +1,119 @@
+    const auto p = posRelToSide(HYPOTENUSE, x, y);
+    const auto rad = std::hypot(p.x, p.y);
+
+    if (rad >= geometryTool->getHeight() / std::sqrt(2.) - 1.15 || p.y > 0) {  // TODO
```
This is duplicating 

https://github.com/xournalpp/xournalpp/blob/fb63352f6d39bd1c5be712e31ff4529bbf241bf5/src/core/view/SetsquareView.cpp#L52-L52

There is probably a better way to do it, but I don't want to include `SetsquareView.h` in `SetsquareController.cpp`
Any idea?

bhennion replied:
From what I could gather in a glance, you put all the lengths constants in SetSquareView. I think they should be part of the model (to which both the view and the controller have access). For instance DISTANCE_SEMICIRCLE_FROM_LEGS could be a member `static constexpr` in SetSquare. You could also imaging a static function radiusFromHeight() in SetSquare.

rolandlo replied:
Done. I think the constants that are only used for rendering should remain in `SetsquareView` though.

rolandlo commented on src/core/control/Control.cpp:
```diff
@@ -646,21 +650,16 @@ void Control::actionPerformed(ActionType type, ActionGroup group, GtkToolButton*
+                this->geometryToolController->getType() != GeometryToolType::SETSQUARE) {
+                resetGeometryTool();
+                makeGeometryTool(GeometryToolType::SETSQUARE);
+                xournal->getViewFor(getCurrentPageNo())->rerenderPage();
```
I removed the page rendering and it still works fine. I had to pass the tool range of the setsquare as the last repaint range to the geometry tool constructor, since otherwise there were artifacts when translating the setsquare using the arrow keys.

rolandlo commented on src/core/control/Control.cpp:
```diff
@@ -646,21 +650,16 @@ void Control::actionPerformed(ActionType type, ActionGroup group, GtkToolButton*
-                // hide setsquare
-                this->win->getXournal()->resetSetsquareView();
+                resetGeometryTool();
+                xournal->getViewFor(getCurrentPageNo())->rerenderPage();
```
Removed as well.

rolandlo commented on src/core/control/Control.cpp:
```diff
@@ -1351,12 +1374,13 @@ void Control::updateDeletePageButton() {
+    if (geometryToolController && doc->indexOf(geometryToolController->getPage()) == pNr) {
+        resetGeometryTool();
     }
+    doc->unlock();
```
Done

rolandlo commented on src/core/control/Control.cpp:
```diff
@@ -1053,6 +1052,30 @@ void Control::actionPerformed(ActionType type, ActionGroup group, GtkToolButton*
+        std::unique_ptr<GeometryToolInputHandler> geometryToolInputHandler =
+                std::make_unique<SetsquareInputHandler>(this->win->getXournal(), geometryToolController.get());
+        geometryToolInputHandler->registerToPool(setsquare->getHandlerPool());
+        auto xournal = GTK_XOURNAL(this->win->getXournal()->getWidget());
```
Ok, done.

rolandlo commented on src/core/control/Control.cpp:
```diff
@@ -1053,6 +1052,30 @@ void Control::actionPerformed(ActionType type, ActionGroup group, GtkToolButton*
+void Control::resetGeometryTool() {
+    this->geometryToolController.reset();
+    this->geometryTool.reset();
+    auto xournal = GTK_XOURNAL(this->win->getXournal()->getWidget());
```
Done

rolandlo commented on src/core/control/GeometryToolController.h:
```diff
@@ -0,0 +1,102 @@
+     * @param x the translation in x-direction (in document coordinates)
+     * @param y the translation in y-direction (in document coordinates)
+     */
+    void move(double x, double y);
```
Ok, I renamed it to `translate`. I used `move` before because it's the same in `EditSelection`.

rolandlo commented on src/core/control/SetsquareController.cpp:
```diff
@@ -0,0 +1,119 @@
+    addStrokeToLayer();
+}
+
+auto SetsquareController::existsEdgeStroke() -> bool { return !(hypotenuseMax == std::numeric_limits<double>::min()); }
```
Done

rolandlo commented on src/core/control/SetsquareController.h:
```diff
@@ -0,0 +1,130 @@
+     * x-coordinates of the point to be drawn (with respect to an unrotated, and untranslated coordinate system) are
+     * saved in the variables hypotenuseMin and hypotenuseMax
+     */
+    double hypotenuseMax = std::numeric_limits<double>::min();
```
Thanks for the catch!

bhennion replied:
Did you also change every other appearance of `std::numeric_limits<double>::min()` in the rest of the code (e.g. when comparing)?

rolandlo replied:
Yes. The only occurence now is in `safe_casts.h` as a comment (which is completely unrelated to this PR).

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+            if (zoomGesturesEnabled) {
+                zoomStart();
+                return true;
+            }
```
There was something wrong when zoom gestures were not enabled. I had not tried that out previously. As soon as you had two fingers on the screen the setsquare would not do anything. Now the setsquare will still rotate when zoom gestures are disabled, it just won't scale.

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    if (event.type == MOTION_EVENT && this->primarySequence) {
+        if (this->primarySequence && this->secondarySequence && zoomGesturesEnabled) {
```
Done

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+        } else if (event.sequence == this->primarySequence) {
+            scrollMotion(event);
+            return true;
+        } else if (this->primarySequence && this->secondarySequence) {
```
Done

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+            scrollMotion(event);
+            return true;
+        } else if (this->primarySequence && this->secondarySequence) {
+            sequenceStart(event);
```
Done

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+            // If secondarySequence is nullptr, this sets primarySequence
+            // to nullptr. If it isn't, then it is now the primary sequence!
+            this->primarySequence = this->secondarySequence;
+            this->secondarySequence = nullptr;
```
Done

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+        } else {
+            this->secondarySequence = nullptr;
+        }
+    }
```
Done

rolandlo commented on src/core/gui/inputdevices/GeometryToolInputHandler.cpp:
```diff
@@ -0,0 +1,335 @@
+
+auto GeometryToolInputHandler::handleKeyboard(InputEvent const& event) -> bool {
+    GdkEvent* gdkEvent = event.sourceEvent;
+    if (gdkEvent->type == GDK_KEY_PRESS) {
```
Done

rolandlo commented:
Applied all reviewer suggestions and rebased on current master, while solving rebase conflicts. Also fixed an issue that occurred when zoom gestures were disabled.

bhennion commented on src/core/control/GeometryToolController.cpp:
```diff
@@ -0,0 +1,122 @@
+        case TOOL_ERASER:
+            stroke->setToolType(StrokeTool::ERASER);
+            break;
+        default:  // Should not occur
+            break;
```
Maybe print a warning here

bhennion commented on src/core/model/GeometryTool.cpp:
```diff
@@ -0,0 +1,57 @@
+
+void GeometryTool::setStroke(Stroke* s) { this->stroke = s; }
+
+auto GeometryTool::computeRepaintRange(Range& rg) const -> Range {
```
Are you purposefully modifying the parameter Range here?
```suggestion
auto GeometryTool::computeRepaintRange(Range rg) const -> Range {
```

bhennion reviewed (approved):
Two nitpicks. Otherwise, LGTM!

rolandlo commented:
> Two nitpicks. Otherwise, LGTM!

Thanks for the review once more! I have fixed the two issues. Now I will clean up the commit history a bit (without changing the total diff).

rolandlo commented:
Cleaned up the commit history a bit (with no functional changes) and rebased on current master. 
Merging in 48 hours if no objections are raised.
